src/Shared/CommunicationsUtilities.cs

```diff
@@ -365,39 +318,49 @@ internal static void WriteLongForHandshake(this PipeStream stream, long value)
                 Array.Reverse(bytes);
             }
 
-            ErrorUtilities.VerifyThrow(bytes.Length == 8, "Long should be 8 bytes");
+            ErrorUtilities.VerifyThrow(bytes.Length == 4, "Int should be 4 bytes");
 
             stream.Write(bytes, 0, bytes.Length);
         }
 
-        /// <summary>
-        /// Extension method to read a series of bytes from a stream
-        /// </summary>
-        internal static long ReadLongForHandshake(this PipeStream stream
+        internal static void ReadEndOfHandshakeSignal(this PipeStream stream, bool isProvider
 #if NETCOREAPP2_1 || MONO
-            , int handshakeReadTimeout
+            , int timeout
 #endif
             )
         {
-            return stream.ReadLongForHandshake((byte[])null, 0
+            // Accept only the first byte of the EndOfHandshakeSignal
+            int valueRead = stream.ReadIntForHandshake(0x00
 #if NETCOREAPP2_1 || MONO
-                , handshakeReadTimeout
+            , timeout
 #endif
                 );
+
+            if (valueRead != EndOfHandshakeSignal)
+            {
+                if (isProvider)
+                {
+                    CommunicationsUtilities.Trace("Handshake failed on part {0}. Probably the client is a different MSBuild build.", valueRead);
+                }
+                else
+                {
+                    CommunicationsUtilities.Trace("Expected end of handshake signal but received {0}. Probably the host is a different MSBuild build.", valueRead);
+                }
+                throw new InvalidOperationException();
+            }
         }
 
         /// <summary>
         /// Extension method to read a series of bytes from a stream.
         /// If specified, leading byte matches one in the supplied array if any, returns rejection byte and throws IOException.
         /// </summary>
-        internal static long ReadLongForHandshake(this PipeStream stream, byte[] leadingBytesToReject,
-            byte rejectionByteToReturn
+        internal static int ReadIntForHandshake(this PipeStream stream, int byteToAccept
```

Comment:
nit: Typing `byteToAccept` as `int?` would make it slightly more readable (no special meaning of 0x00). Also, the caller already handles the general case of a handshake component mismatch so maybe it should handle the leading byte mismatch as well? It could be done through a helper on `struct Handshake` to keep the two sides together - OR'ing `options` with 0x01... and checking that it is there.

Reply:
Never mind, the leading byte needs to be handled here so we don't wait for more bytes on mismatch 🤦